### PR TITLE
Optimize deriveshaOrig

### PR DESCRIPTION
**Pull request by hqjang-pepper**
## Proposed changes

This PR contains minor deriveSha optimization.
- AppendUint64 appends the RLP encoding of i to b, and returns the resulting slice. (an rlp integer is known to never require more than 9 bytes total). This function makes encoding more faster by using bit shifting
- change some buffer reseting logic. 

Plus. Added some benchmark testcode.

related Ethereum PR : https://github.com/ethereum/go-ethereum/pull/21728

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by kjhman21** [comment]
@hqjang-pepper Can you share the performance result of the benchmark here as a comment?

**Comment by hqjang-pepper** [comment]
Tested with 100, 1000, 4000 txs, using StackTrie

```
Before this PR :
BenchmarkDeriveShaOrig/1000test-stackTrie-10         	    7971	    153447 ns/op
BenchmarkDeriveShaOrig/1000test-stackTrie-10         	     793	   1499944 ns/op
BenchmarkDeriveShaOrig/4000test-stackTrie-10         	     195	   5928197 ns/op

After this PR :
BenchmarkDeriveShaOrig/100test-stackTrie-10         	    7995	    149334 ns/op. -  2.7% faster
BenchmarkDeriveShaOrig/1000test-stackTrie-10         	     788	   1488989 ns/op. -  0.7% faster
BenchmarkDeriveShaOrig/4000test-stackTrie-10         	     199	   5891009 ns/op. -  0.6% faster

```
2.7%

**Comment by hqjang-pepper** [comment]
Tested with 100, 1000, 4000 txs, using StackTrie

Before this PR : 
BenchmarkDeriveShaOrig/100test-stackTrie-10         	    7977	    148040 ns/op
BenchmarkDeriveShaOrig/1000test-stackTrie-10         	     744	   1533285 ns/op
BenchmarkDeriveShaOrig/1000test-stackTrie-10         	     194	   5926051 ns/op


After this PR :
BenchmarkDeriveShaOrig/100test-stackTrie-10         	    8449	    140057 ns/op  -  5.7% faster
BenchmarkDeriveShaOrig/1000test-stackTrie-10         	     836	   1441072 ns/op. -  6.4% faster
BenchmarkDeriveShaOrig/4000test-stackTrie-10         	     210	   5555414 ns/op. -  6.6% faster

File: storage/statedb/derive_sha.go [hunk 1]
```diff
@@ -33,26 +31,23 @@ type DeriveShaOrig struct{}
...
 	}
 	for i := 0x80; i < list.Len(); i++ {
-		keybuf.Reset()
-		rlp.Encode(keybuf, uint(i))
-		trie.Update(keybuf.Bytes(), list.GetRlp(i))
+		buf = rlp.AppendUint64(buf[:0], uint64(i))
+		trie.Update(buf, list.GetRlp(i))
 	}
 	return trie.Hash()
 }
```
Comment (hyunsooda) [inline storage/statedb/derive_sha.go:53]: Seems conflict. You did migrate it to `types` package in the preceding PR, but here not reflected.

File: tests/tx_root_test.go [hunk 2]
```diff
@@ -33,7 +35,7 @@ func BenchmarkDeriveSha(b *testing.B) {
 		"Concat": types.DeriveShaConcat{},
 	}
 
-	NTS := []int{1000}
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply tests/tx_root_test.go]: Rollbacked

File: tests/tx_root_test.go [hunk 3]
```diff
@@ -81,9 +104,32 @@ func benchDeriveSha(b *testing.B, numTransactions, numValidators int, sha types.
 
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
-		hash := sha.DeriveSha(txs)
-		if testing.Verbose() {
-			fmt.Printf("[%d] txhash = %s\n", i, hash.Hex())
+		sha.DeriveSha(txs)
+		//if testing.Verbose() {
+		//	fmt.Printf("[%d] txhash = %s\n", i, hash.Hex())
+		//}
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply tests/tx_root_test.go]: Too many unnecessary lines are printed, hard to figure out the test result.
So I deleted this line.

File: storage/statedb/derive_sha.go (hunk 1 above)
Comment (hqjang-pepper) [reply storage/statedb/derive_sha.go:53]: Yes that's right. For easier review I didn't reflected this change in the previous PR.
I'll resolve conflict after this PR is merged

Comment (blukat29) [reply storage/statedb/derive_sha.go:53]: I agree that this PR should be merged first.

**Review by blukat29 — APPROVED** [review]

**Review by kjhman21 — APPROVED** [review]
LGTM

**Review by hyunsooda — APPROVED** [review]

File: storage/statedb/derive_sha.go (hunk 1 above)
Comment (hyunsooda) [reply storage/statedb/derive_sha.go:53]: Okay. thanks.

**Review by jeongkyun-oh — APPROVED** [review]

**Pull request merged**